File: src/hash.ts
import { createHash, timingSafeEqual } from "node:crypto";

export function hashPassword(pw: string): string {
  return createHash("sha256").update(pw).digest("hex");
}

/** Check a plaintext password against a stored hash: verify(password, storedHash). */
export function verify(pw: string, hash: string): boolean {
  const a = Buffer.from(hashPassword(pw), "hex");
  const b = Buffer.from(hash, "hex");
  return a.length === b.length && timingSafeEqual(a, b);
}

File: src/hash.py
from __future__ import annotations

import hashlib
import hmac


def hash_password(pw: str) -> str:
    return hashlib.sha256(pw.encode()).hexdigest()


def verify(pw: str, stored_hash: str) -> bool:
    """Check a plaintext password against a stored hash: verify(password, stored_hash)."""
    a = bytes.fromhex(hash_password(pw))
    b = bytes.fromhex(stored_hash)
    return len(a) == len(b) and hmac.compare_digest(a, b)
